### PR TITLE
`sage.rings.function_field`: Modularization fixes

### PR DESCRIPTION
**Pull request by mkoeppe**
### 📚 Description
- We split out `.derivations` (which uses modules) from `.maps` (which doesn't). 
- We separate the implementation of rational function fields (which do not need Singular for Gröbner basis calculations) from the implementation of more complicated function fields (which do).
- We move some imports into methods.

This is for modularization purposes:
- Part of #29705
 

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [ ] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### ⌛ Dependencies

- Depends on #35237 (for the feature `sage.libs.singular`)
- Depends on #35119 (to preempt merge conflicts)

### CODE REVIEW

**Comment by kwankyu** [comment]
This PR makes me think whether the direction and the level of the modularization efforts is reasonable.

The branch adds one of `# optional - M` tags to *every* doctests depending on the base fields over which function fields are defined, where `M` is one of `sage.libs.singular`, `sage.libs.pari`, and `sage.rings.number_field`, and also `sage.modules` for derivations. In the future, no doctest in the function fields module will go without one of these tags since a function field needs a base field. This is very fine modularization that I didn't expect.

If the function fields module is dependent on all of these `M`s, we do not need these tags, and one can add a new doctest without worrying which one of those tags should be added with the doctest. 

So a basic question is whether some users of the function fields module will need only, say, finite fields as their base fields and want to install only `sage.libs.singular` along with the function fields module, and thus he/she can save some of disk space and install time. I guess there can be such users for some critical mission, but most users (including me) will be satisfied with time and space saved by much coarse modularization like, say, sage with full function fields module but without manifolds and optimization modules.

If we can choose the level of the whole modularization efforts, I think we need to decide this on sage-devel, once and for all, before proceeding.

**Comment by mkoeppe** [comment]
> In the future, no doctest in the function fields module will go without one of these tags since a function field needs a base field.

No, that's not true. The basic rings ZZ and QQ are always available.

**Comment by kwankyu** [comment]
Ah, I see. If the function field is an extension of rational function field over `QQ`, the doctest needs `sage.libs.singular`. So to decide which tag is needed, one needs to know the implementation detail.

**Comment by mkoeppe** [comment]
Yes, these dependencies are not always obvious. To some extent these optional tags provide the necessary information for users who need to deploy the modularized distributions.
I would also expect that there may be a few ways how the hard dependencies on some of the libraries can be avoided. For example, Singular is involved whenever a Gröbner basis must be computed. But it's called even for very very simple inputs.

**Comment by mkoeppe** [comment]
For modules where all doctests depend on a feature, I have been using the file-level annotation. `# sage.doctest: optional - M`.
But most modules mix tests that require no additional features with some that do require additional features.

**Comment by kwankyu** [comment]
All function fields, which are not simple rational function fields, use multivariate polynomial rings, which is provided by Singular. As a user of the function fields module, Singular is a prerequisite, not an additional feature.

**Comment by mkoeppe** [comment]
That's true, currently; but soon we will be able to provide multivariate polynomials using FLINT instead.

**Comment by kwankyu** [comment]
Then those optional tags will need to be updated? This is a drawback of technical dependency added to doctests?

**Comment by mkoeppe** [comment]
Yes, when the underlying library for a functionality changes, some of these annotations will need updating.

Currently for example, the tag `sage.libs.pari` is on all tests that construct a finite field; but it is also on all tests that require factoring & prime number ranges. I haven't attempted to distinguish these two cases [there are probably more] using semantically higher level tags. 

On the other hand, a tag like `sage.rings.number_field` will not need to be updated when we reimplement number fields.

**Comment by kwankyu** [comment]
Back to my main point: We can just make the distribution package that installs the function fields module dependent on those technical features. Why not? The same question will be raised for every high level mathematical module when most of its doctests would be laced with lots of repetitive `optional` tags.

**Comment by mkoeppe** [comment]
> Why not?

Because the ordinary rational functions also go through the same code, and they are a crucial of setting up the categories for the basic rings.

**Comment by mkoeppe** [comment]
In some cases, after a mild reorganization it may be possible to replace lots of `# optional` tags with a file-level annotation.
For example `FunctionFieldMaximalOrderInfinite_polymod` currently only has doctests using finite fields, not rationals [I have not checked whether that's by design]. So if this were moved to a separate file, we would have much fewer tag.

**Comment by kwankyu** [comment]
It would be possible to separate rational functions code from the extensions (of rational function fields) code. Then rational functions module could be included in a small distribution, and the extensions module would go into a bigger distribution.

**Comment by mkoeppe** [comment]
Yes, that could be a nice solution. I really only put the general function fields in the small distribution **sagemath-categories** to solve the problem with the missing rational functions.

**Comment by mkoeppe** [comment]
Or, even simpler, we could extract the tests for rational functions and move these to a test-only module.

**Comment by kwankyu** [comment]
The tests are also examples. How could they be separated from the methods code?

**Comment by kwankyu** [comment]
>  For example `FunctionFieldMaximalOrderInfinite_polymod` currently only has doctests using finite fields, not rationals [I have not checked whether that's by design]. 

That is not by design. The author (me) just happened to use the class only for finite fields.

**Comment by kwankyu** [comment]
> Yes, that could be a nice solution.

Okay. I will try that. It may not be difficult...

**Comment by mkoeppe** [comment]
In any case, if we were to move the `_polymod` implementations into a separate file, then I can drop the `# optional - modules`

**Comment by kwankyu** [comment]
> Yes, that could be a nice solution. 

It is not easy.

> I really only put the general function fields in the small distribution **sagemath-categories** (in the version of #35095) to solve the problem with the missing rational functions.

I don't see a direct use of function fields in the categories module. Perhaps indirectly used for fractions of polynomials... 

What happens if function fields module is excluded from the distribution **sagemath-categories**? Lots of doctest failures?

**Comment by mkoeppe** [comment]
> if we were to move the `_polymod` implementations into a separate file, then I can drop the `# optional - modules`

I've split the module `.order` along these lines. If this is going in a good direction, I can do the same with `.ideal`.

**Comment by kwankyu** [comment]
Is the function fields module included in `sagemath-categories`? 

I don't see it in `pkgs/sagemath-categories/MANIFEST.in`...

**Comment by kwankyu** [comment]
If sagemath-polyhedra needs the (rational) functions fields module, why not put it to sagemath-polyhedra instead of sagemath-categories?  

Putting the function fields module to sagemath-categories only for some technical reason seems not a good idea.

**Comment by mkoeppe** [comment]
All of what's in and what's out is preliminary at this point, of course. I'm trying to make sagemath-categories sufficiently large so that it gets good test coverage for the coercion system, including pushouts. I think based on the errors that I have seen that rational function fields are important for this (but at the moment I can't really substantiate or quantify this). Some more experiments are needed.

**Comment by kwankyu** [comment]
I am following the pattern you suggested to split out rational function fields module, which needs to be included in sagemath-categories. 

Other than that, the whole function fields module, I think, should be dependent on singular, pari, and whatever library that it needs, and thus there's no need to add tags for them. Do you agree?

**Comment by mkoeppe** [comment]
Yes, this sounds good.

**Comment by kwankyu** [comment]
I made a PR based on your branch.

**Comment by mkoeppe** [comment]
> Perhaps before you update optional tags, do we need to change or move around doctests (and documentation) in, say, `order.py`? Perhaps this will be installed by sagemath-categories as part of the rational function fields module.

Yes, `order.py` and `order_rational.py` will be shipped in **sagemath-categories** (in the version of #32432). 

>  Then could it have doctests about non-rational function fields? 

Yes. As of this branch, it has such doctests already. They are currently marked `sage.libs.singular`.

> Such doctests should have, say, `# optional - sage.rings.function_field?`

Yes, that could be a possible new doctest tag; so far I was just using `sage.libs.singular`, with a (to-be-defined) distribution **sagemath-singular** in mind that would just ship everything that needs Singular. 

But I think your suggestion `sage.rings.function_field` is better, so I'll change the existing `# optional` tags accordingly.

**Comment by mkoeppe** [comment]
> I think your suggestion `sage.rings.function_field` is better, so I'll change the existing `# optional` tags accordingly.

I've committed the first changes in this direction.

**Comment by mkoeppe** [comment]
I've merged this into #32432, it seems to work fine (there, in e3297cfb33, I'm excluding the `..._polymod` modules)

File: src/sage/rings/function_field/function_field.py [hunk 1]
```diff
@@ -278,7 +274,7 @@ class FunctionField(Field):
         sage: K
         Rational function field in x over Rational Field
     """
-    _differentials_space = DifferentialsSpace
+    _differentials_space = LazyImport('sage.rings.function_field.differential', 'DifferentialsSpace')
```
Comment (kwankyu) [inline src/sage/rings/function_field/function_field.py:277]: Why is this necessary?

**Comment by kwankyu** [comment]
As we discussed, I would like `sage.libs.pari` to be included to `sage.rings.function_field`, or otherwise would prefer the name `sage.rings.finite_field` (or `sage.rings.finite_rings`) instead of `sage.libs.pari` to hide implementation details. 

But I fear that this change disrupts other parts of modularization, and moreover this PR is already big. So I would not insist it for this PR. 

Otherwise looks good to me. Would you just reply to my question on `LazyImport` above?

**Comment by mkoeppe** [comment]
> Would you just reply to my question on `LazyImport` above?

Sorry, which question?

**Comment by mkoeppe** [comment]
> As we discussed, I would like `sage.libs.pari` to be included to `sage.rings.function_field`, or otherwise would prefer the name `sage.rings.finite_field` (or `sage.rings.finite_rings`) instead of `sage.libs.pari` to hide implementation details.

Sure, either of these two solutions would work, and it's best to do it right here on the PR. I'll look into this tomorrow.

File: src/sage/rings/function_field/function_field.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/rings/function_field/function_field.py:277]: In sagemath-categories, I include (basic) rings but not modules; and `DifferentialsSpace` is a module

**Comment by mkoeppe** [comment]
> I would like `sage.libs.pari` to be included to `sage.rings.function_field`, or otherwise would prefer the name `sage.rings.finite_field` (or `sage.rings.finite_rings`) instead of `sage.libs.pari` to hide implementation details.

I went with the second option because I think the characteristic 0 tests do not need pari (but I can't test this at the moment). The feature is named `sage.rings.finite_rings` (not `...finite_field` because I'm using actual module names in features with `.`)

File: src/sage/rings/function_field/maps.py [hunk 2]
```diff
@@ -45,1081 +35,39 @@
...
+lazy_import("sage.rings.function_field.derivations", (
+    "FunctionFieldDerivation",
+    "FunctionFieldDerivation_rational",
+    "FunctionFieldDerivation_separable",
+    "FunctionFieldDerivation_inseparable",
+    "FunctionFieldHigherDerivation",
+    "RationalFunctionFieldHigherDerivation_global",
+    "FunctionFieldHigherDerivation_global",
+    "FunctionFieldHigherDerivation_char_zero",
+), deprecation=35230)
```
Comment (kwankyu) [inline src/sage/rings/function_field/maps.py:64]: I don't think that these deprecations are necessary. They are internal modules, and not intended to be used directly by users.

File: src/sage/rings/function_field/order.py [hunk 3]
```diff
@@ -94,51 +94,35 @@
...
-
-from .hermite_form_polynomial import reversed_hermite_form
+lazy_import('sage.rings.function_field.order_basis',
+            ['FunctionFieldOrder_basis', 'FunctionFieldOrderInfinite_basis'])
+lazy_import('sage.rings.function_field.order_rational',
+            ['FunctionFieldMaximalOrder_rational', 'FunctionFieldMaximalOrderInfinite_rational'])
+lazy_import('sage.rings.function_field.order_polymod',
+            ['FunctionFieldMaximalOrder_polymod', 'FunctionFieldMaximalOrderInfinite_polymod'])
+lazy_import('sage.rings.function_field.order_global',
+            ['FunctionFieldMaximalOrder_global'])
```
Comment (kwankyu) [inline src/sage/rings/function_field/order.py]: Perhaps, these are not necessary. They are internal modules.

**Comment by kwankyu** [comment]
> I went with the second option because I think the characteristic 0 tests do not need pari (but I can't test this at the moment). The feature is named `sage.rings.finite_rings` (not `...finite_field` because I'm using actual module names in features with `.`)

Okay. I agree. Thanks a lot. 

I think hiding implementation details from feature names is important. In general, for maintainability of the modularization of sage library, it should be, ideally, obvious to add which `# optional - M` tags to which doctests. It is best if no tag is necessary. I feel that there should be still more improvement in this direction, though I have no idea at this moment.

File: src/sage/rings/function_field/maps.py (hunk 2 above)
Comment (mkoeppe) [reply src/sage/rings/function_field/maps.py:64]: OK, done.

**Comment by mkoeppe** [comment]
> I think hiding implementation details from feature names is important. In general, for maintainability of the modularization of sage library, it should be, ideally, obvious to add which `# optional - M` tags to which doctests. It is best if no tag is necessary. I feel that there should be still more improvement in this direction, though I have no idea at this moment.

I have added something along these lines to the developer guide.

**Comment by kwankyu** [comment]
Perfect. Thanks.

**Review by kwankyu — APPROVED** [review]

**Comment by mkoeppe** [comment]
Thanks! I also approve your changes, merged in the branch.

**Comment by kwankyu** [comment]
I added missing titles to new files in the PR to your branch

**Comment by kwankyu** [comment]
Nice. Thanks.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
merge conflict

**Comment by mkoeppe** [comment]
Trivial merge, back to positive

**Comment by vbraun** [comment]
```
sage -t --long --warn-long 73.8 --random-seed=123 src/sage/interfaces/r.py  # 3 doctests failed
```

**Comment by saraedum** [comment]
@kwankyu many of the copyright headers here now only mention you as the copyright holder. I think that's not accurate and all the authors of the files from which the code has been copied should be mentioned.

**Comment by kwankyu** [comment]
Right. Sorry about that. I prepared a PR to fix the mistakes.

**Comment by mkoeppe** [comment]
Also apologies from me – apparently I didn't catch it when I reviewed @kwankyu's contributions to this PR
